File: Data Structures & Algorithms/max-water-container/submission-0.rs
impl Solution {
    // we're maximizing min(n[l], n[r]) * (r - l)
    // on each iteration, we move the pointer
    // pointing to the lower bar.
    // so essentially. we're looking for two
    // highest bars but also consider distance between them.
    //
    // the bigger Q: how we ensure we cover all the possible valid cases??
    pub fn max_area(h: Vec<i32>) -> i32 {
        // 2 <= height.length, so fine to len()-1
        let (mut l, mut r) = (0usize, h.len()-1);

        let mut mx = 0;
        while l < r {
            let mut s = 0i32;
            if h[l] < h[r] {
                s = h[l] * (r-l) as i32;
                l += 1;
            } else {
                s = h[r] * (r-l) as i32;
                r = r.saturating_sub(1);
            }
            mx = mx.max(s);
        }

        mx
    }
}

File: Data Structures & Algorithms/median-of-two-sorted-arrays/submission-0.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& a, vector<int>& b) {
        if (a.size() > b.size()) return findMedianSortedArrays(b, a);

        int m = a.size(), n = b.size();
        int half = (m + n + 1) / 2;
        int lo = 0, hi = m;

        while (lo <= hi) {
            int i = lo + (hi - lo) / 2;
            int j = half - i;

            long aL = (i == 0) ? LONG_MIN : a[i-1];
            long aR = (i == m) ? LONG_MAX : a[i];
            long bL = (j == 0) ? LONG_MIN : b[j-1];
            long bR = (j == n) ? LONG_MAX : b[j];

            if (aL <= bR && bL <= aR) {
                if ((m + n) % 2) return max(aL, bL);
                return (max(aL, bL) + min(aR, bR)) / 2.0;
            }
            if (aL > bR) hi = i - 1;
            else         lo = i + 1;
        }
        return 0.0;
    }
};